### ISSUE TITLE
sending emails

### ISSUE THREAD

**Issue by sylwester-** [brefphp/bref#612]
hello,

First of all, fantastic platform! Very good documentation. One question that I would have is around sending emails. I've been looking into this for few days but can't seem to find a way of sending emails from laravel using Notifications or even simply sending emails. Lambda sits within VPC.

Is there documentation that I could follow that is more detailed than this: #571

**Comment by victormacko** [comment]
So one downside of putting your Lambda inside a VPC is that it's prevented from accessing the internet (including AWS's own email service) ... best approach if you don't actually need to be inside a VPC is to move your lambda outside of the VPC.
If you do need the VPC otherwise, setup a NAT Instance (little EC2 to act as your gateway ~$5 per month) or a NAT Gateway (managed gateway - ~$40 a month from memory) to get internet access sorted.

Once that's done - just use SES to send email.
There are other approaches of course (as in that issue you linked to) if you have more time to spend and are certain that you don't need to access any other APIs / services.

**Comment by sylwester-** [comment]
@victormacko  - thank you for a swift response. I'm currently looking into creating my own laravel mail driver that would use lambda to send emails. Do you see any risks in that approach? anything I would need to watch out for? Based on the #571 bref lambda should be able to communicate with my working "send email" lambda.

**Comment by victormacko** [comment]
Maybe just make sure your VPC lambda can actually talk to your mailing lambda (or whatever you're going to be using to trigger your mailing lambda) -- SNS is a good go-between, being async.

I could be wrong, but the issue you referenced uses SNS ... which in-turn uses Amazon VPC endpoints for the VPC comms ... which in-turn is powered by AWS PrivateLink (which has a per-hour billing = about $7.20 per month per AZ per endpoint (eg. SNS = 1 endpoint from what I can see))

I'd suggest checking out the Bref Slack channel for advice like this - https://bref.sh/docs/community.html ... the community there are super knowledgable & much faster at responding there.
... if you do find a better way, let us know!

**Comment by sylwester-** [comment]
@victormacko - might be a silly question but how do you actually enable that communication between two lambdas? I have code that works great locally, invokes the lambda and lambda sends an email, but obviously this doesn't work when I push it to my lambda dev environment. A played with policies and roles in IAM but no luck

**Comment by victormacko** [comment]
I've not actually setup comms between two VPC lambdas (without external access before) - I gather you'd have some sort of internal DNS to use & trigger them via HTTP calls.

If you're referring to your email setup and if it triggered from a Lambda within a VPC (without a NAT gateway/instance) you're going to be a little stuck, not being able to access external SNS / HTTP apis / etc ... you could always try just writing your requests to S3 (eg. as JSON files), and then using S3 events to process the request ... i'd prob not advise it though if you're doing lots of these (pretty sure the PUT/POST pricing on S3 would get pricy after a bit). DynamoDB I think "might" be accessible within a VPC and you can use those events in the same way - but check the docs on that first.

Honestly if cost isn't a big issue, i'd just bite the bullet and setup a NAT instance/gateway ... in my experience, once this project is done, the next one will start needing more internet access, and so-on... to the point all the "tricks" to get around it just aren't worth it.

**Comment by sylwester-** [comment]
@victormacko  - finally managed to make it work. I went with the DynamoDB approach and it worked very well. Had to create a VPC endpoint for dynamodb but that is very simple to do. 
I ended up with this setup: 

laravel lambda with custom mail driver that saved email to dynamodb via VPC endpoint. when new record is added to dynamodb table, "send email lambda" is triggered and sending is done via SES. 

Happy to provide more details if someone needs it.
